### PR TITLE
Remove react-native for 0.73.0

### PR DESCRIPTION
**Pull request by lunaleaps**
React Native ships its own types since 0.71 and plans to deprecate DT types for 0.73: https://reactnative.dev/blog/2023/01/03/typescript-first

This PR removes the types via
`pnpm run not-needed react-native 0.73.0`

- [x] Delete the package's directory.
- [x] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by jakebailey** [comment]
At the moment, this PR is most likely wrong because it's not also updating all of the other DT packages which depend on `react-native`; those would all need to be updated too. It's likely that the CI won't be able to test this yet until `notNeededPackages` is valid and `react-native` is added to https://github.com/microsoft/DefinitelyTyped-tools/blob/master/packages/definitions-parser/allowedPackageJsonDependencies.txt.

You may want to do the latter now, then go and update all of other packages within the repo to point to the correct react-native and see what happens,

**Comment by lunaleaps** [comment]
@jakebailey Thanks!
Here's the PR to update allowed packages: https://github.com/microsoft/DefinitelyTyped-tools/pull/839/files

**Comment by lunaleaps** [comment]
Looking at the error, it looks like this might also be blocked on us releasing react-native 0.73. I'm not sure we need to update the other libraries that use react-native as they should be still able to reference the npm packages for older versions. 

Let's wait until 0.73.0 is out, then we can revisit this.

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## react-native-calendars ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-native-calendars@latest/))
was missing the following properties:
1. AgendaList
1. asCalendarConsumer
## react-native-countdown-component ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-native-countdown-component@latest/))
was missing the following properties:
1. CountDown
## react-native-indicators ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-native-indicators@latest/))
was missing the following properties:
1. Indicator
## react-native-modals ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-native-modals@latest/))
was missing the following properties:
1. BackdropProps
1. ModalContentProps
1. ModalTitleProps
1. ModalButtonProps
1. ModalFooterActionList

<details>
<summary>as well as these 4 other properties...</summary>
<p>ModalFooterProps, ModalProps, SwipeDirection, DragEvent</p>
</details>

## react-native-popup-dialog ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-native-popup-dialog@latest/))
was missing the following properties:
1. OverlayProps
1. DialogContentProps
1. DialogTitleProps
1. DialogButtonProps
1. DialogActionListProps

<details>
<summary>as well as these 4 other properties...</summary>
<p>DialogProps, Animation, DialogActionList, Dialog</p>
</details>

## react-native-signature-capture ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-native-signature-capture@latest/))
was missing the following properties:
1. The declaration doesn't match the JavaScript module 'react-native-signature-capture'. Reason:
The JavaScript module can be called or constructed, but the declaration module cannot.

The most common way to resolve this error is to use 'export =' syntax.
To learn more about 'export =' syntax, see https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require.
1. propTypes
## react-native-text-input-mask ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-native-text-input-mask@latest/))
was missing the following properties:
1. mask
1. unmask
1. setMask
## react-native-webrtc ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-native-webrtc@latest/))
was missing the following properties:
1. permissions
## react-primitives ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-primitives@latest/))
was missing the following properties:
1. inject
## react-router-navigation ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-router-navigation@latest/))
was missing the following properties:
1. DefaultNavigationRenderer
1. DefaultTabsRenderer
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 81befa82831b1026831ac2f6ec0d1515ee650b51
</p>

File: types/cavy/package.json
Comment (jakebailey) [inline types/cavy/package.json:13]: I believe that these should all now depend on `react-native` directly, or, leave this as-is as it'll automatically pull in the shim?

Comment (lunaleaps) [reply types/cavy/package.json:13]: Hmm I guess for testing we should probably point it to `react-native`, but I'm concerned about potential breaking changes for new types if we continue with `*` -- `react-native` doesn't truly follow semver so every minor may contain breaking changes. 

In the past when I cut a new minor version of `types/react-native`, if there was a package that broke with the types, I'd fix it at the last version, that's my inclination here. By that logic, this should be `"~0.72"`, I can update to that instead. What do you think @jakebailey?

Comment (jakebailey) [reply types/cavy/package.json:13]: Pinning will mean that anyone who installs any of these packages will get multiple copies of `react-native` types in their install without a way to fix that, besides using package manager overrides to force overwrite these. If these packages don't specify a range of `react-native` themselves, then I would think we wouldn't try and pin them here unless the break was unavoidable. If you've been pinning them back, then maybe that's okay and precedented, but it just seems like a bad time for downstream users.

`^0.72` is equivalent to `~0.72` as the major version is a zero, so there's no difference there.

Comment (jakebailey) [reply types/cavy/package.json]: @andrewbranch @sandersn do you have any notion of whether or not old packages should depend on `@types/react-native` or `react-native` itself right away? The former may be better just because the deprecation shim will make sure the real thing is installed... We unfortunately can't really test with the shim locally, though.

**Comment by jakebailey** [comment]
Another thing this PR may be missing is the removal of some of the weird overrides / hacks we did to make react-native work. For example, see https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/react-native-calendar-picker/tsconfig.json

**Comment by jmfirth** [comment]
Was pinged presumably due to changes to `react-native-drawer-layout` definitions.  No issues with changes, and they seem to build fine in CI pipelines: https://dev.azure.com/definitelytyped/DefinitelyTyped/_build/results?buildId=170699&view=logs&j=8635df7d-ecde-52a7-c7a8-84b998e35690&t=db23287e-6ac1-50dc-0ddb-ad13e9f6e533

* Processing start: https://dev.azure.com/definitelytyped/DefinitelyTyped/_build/results?buildId=170699&view=logs&j=8635df7d-ecde-52a7-c7a8-84b998e35690&t=db23287e-6ac1-50dc-0ddb-ad13e9f6e533&l=686
* Processing end successfully: https://dev.azure.com/definitelytyped/DefinitelyTyped/_build/results?buildId=170699&view=logs&j=8635df7d-ecde-52a7-c7a8-84b998e35690&t=db23287e-6ac1-50dc-0ddb-ad13e9f6e533&l=689

File: types/cavy/package.json
Comment (andrewbranch) [reply types/cavy/package.json]: react-router has been shipping its own types for a while now, which cause conflicts with `@types/react`, so it seems like it would be better to get `@types/react` out of the picture as often as possible.

**Comment by jakebailey** [comment]
That error seems to be unrelated and is a non-npm package that has started to exist or something...

**Comment by lunaleaps** [comment]
@jakebailey I removed the `paths` field where relevant, is that the hack you're referring to? 

The CI error stops all remaining tests and I also can't reproduce it locally. Should I wait on its fix or can we go ahead with landing?

**Comment by jakebailey** [comment]
Can you merge from master? It passes there, but I'm not sure why this PR isn't passing as GHA runs on a merge commit between the two.

**Comment by jakebailey** [comment]
Ignore the above; the weird failure is something I broke but have a fix for in https://github.com/microsoft/DefinitelyTyped-tools/pull/876.

**Comment by lunaleaps** [comment]
@jakebailey great! thanks for fixing, will I need to rebase after that lands or just retry the CI?

**Comment by jakebailey** [comment]
Just retry (close reopen), but pulling main would work too just to trigger it

**Comment by jakebailey** [comment]
Largely this looks fine, though I am still unsure if we should be depending on `@types/react-native@*`, `@types/react-native@^0.72`, or `react-native@*`.

**Comment by andrewbranch** [comment]
I think it’s `react-native` at my suggestion:

> react-router has been shipping its own types for a while now, which cause conflicts with `@types/react-native`, so it seems like it would be better to get `@types/react-native` out of the picture as often as possible.

But I could be wrong in my reasoning.

**Comment by lunaleaps** [comment]
Hmm okay I've revisited some older conversations: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63276#issuecomment-1346608238 and we don't want users of DT libraries to pull in another copy of `react-native`, especially since they're marked `*`, which may differ from their project version of `react-native`.

Ideally we could mark `react-native` as something like `peerDependenciesMeta` https://docs.npmjs.com/cli/v7/configuring-npm/package-json#peerdependenciesmeta but that kinda breaks how we'd test these.

**Comment by jakebailey** [comment]
> Hmm okay I've revisited some older conversations: [#63276 (comment)](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63276#issuecomment-1346608238) and we don't want users of DT libraries to pull in another copy of `react-native`, especially since they're marked `*`, which may differ from user project versions of `react-native`.

Theoretically, if we point at `@types/react-native` and then `notNeededPackages` publishes the shim that points to `react-native`, that condition would be met, but I don't know if that's actually better.

Realistically, we can try one method and see how it goes, and then switch to another method later.

DT does not yet support peer dependencies, so that isn't an option quite yet.

**Comment by lunaleaps** [comment]
> Theoretically, if we point at @types/react-native and then notNeededPackages publishes the shim that points to react-native, that condition would be met, but I don't know if that's actually better.

@jakebailey So I may not understand the shim process here but when I try installing the deprecated version of `slice-ansi` which deprecated their types: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/67240, I see that the replacement library is installed in my `node_modules` as its specified as dependency in the shim. Is there a way to opt out of that?

**Comment by jakebailey** [comment]
Yes, in general by deleting the package off of DT without adding it to not needed.

**Comment by lunaleaps** [comment]
Sounds good, I'll update this to point to `types/react-native: *` and then remove it from the `notNeeded` json file.

> Realistically, we can try one method and see how it goes, and then switch to another method later. It's all undoable.
 
Great! I guess it wouldn't be an issue because we `*` deps, and we could publish a new shim version 0.73.1 if we had to redo?

**Comment by jakebailey** [comment]
I don't understand how pointing to `*` and not publishing the shim relate; if you do that that's not different than pinning back at 0.72...

**Comment by lunaleaps** [comment]
@jakebailey Sorry, I was confused by your statement 

> Yes, in general by deleting the package off of DT without adding it to not needed.

so to confirm, to publish the shim, you have to add `react-native` to `notNeededPackages.json`? 
I want to ideally publish the shim but have not have the shim have a dependency on `react-native` -- is that possible? 

The idea there is that any library that installs the latest dep on `types/react-native` gets a deprecated package that simply provides the warning that `react-native` now ships its own types but does not actually install `react-native` for you.

**Comment by jakebailey** [comment]
Yes, adding to not-needed is what causes the shim to be published.

There's no way to do that currently; that would cause anyone who updates their types to stop compiling altogether because no new types were brought in to replace them.

The two removal options are to either follow the README guide and publish the shim, or to just delete everything and not make a shim. Note that if `react-native` is present and has types, `@types/react-native` will be ignored.

But I think I'm just creating more confusion with my inability to choose; I'd probably just do what Andrew says...

**Comment by lunaleaps** [comment]
@jakebailey Okay then, I've also talked to relevant folks on React Native side and I think we're okay with this approach. If it becomes a larger issue where users have duplicate react-native packages we can look into shipping a types-only package from our scope.

This PR should be in order then -- is there else blocking approval? 

Overall, I think the DT model of deprecation doesn't work with global packages like react-native -- there was some mention of the previous proposal: https://github.com/microsoft/DefinitelyTyped-tools/issues/433

**Review by andrewbranch — APPROVED** [review]

**Comment by lunaleaps** [comment]
ready to merge

**Comment by andrewbranch** [comment]
@typescript-bot ahem

**Comment by kylebake** [comment]
I've seen it mentioned a few times in the comments here but this change caused our repo to pull in the latest version of react-native alongside our current running `0.71.7` version. Is there a reason these type packages need to pull in the latest version of react native as a dependency? Not sure if there's some extra config that we're missing but I can't seem to find anything

**Comment by jakebailey** [comment]
They don't pull in latest, they pull in `*`. It's up to the package manager to decide which version to use but that specifier means "any version". I guess some package managers may try and bump higher but we didn't write `latest`.

Realistically the only "fix" I can imagine is to switch DT to peer deps, but that has its own challenges. (And wouldn't fix this package now that it's off of DT)

**Comment by NickGerleman** [comment]
DefinitelyTyped doesn't allow doing the correct thing wrt peerDependencies. See https://github.com/microsoft/DefinitelyTyped-tools/issues/433

So, under the current model, a different copy of RN can get pulled in (just like a wrong version of `@types/react-native` could get pulled in before). It requires workarounds like resolutions, or lockfile surgery, when using libraries that don't provide their own types.

The main difference after the change is that this now brings in a second whole copy of RN instead of duplicate types. We have been under the impression, that under properly configured repos, this shouldn't lead to the possibility of resolving the wrong package at runtime (from bundler), but as before, we might get wrong types. If that wasn't the case in practice, I think we were considering publishing a types only package, to at least make that vector safer.

@kylebake what error did you see? Was it related to typechecking, or were multiple copies of RN included in the bundle?

**Comment by kylebake** [comment]
> @kylebake what error did you see? Was it related to typechecking, or were multiple copies of RN included in the bundle?

The full error was: 

```
TypeError: options.shouldAddToIgnoreList is not a function
    at getSourceMapInfo (/Users/kylebaker/repos/beam-brush-unified/node_modules/metro/src/DeltaBundler/Serializers/helpers/getSourceMapInfo.js:18:24)
    at processNextModule (/Users/kylebaker/repos/beam-brush-unified/node_modules/metro/src/DeltaBundler/Serializers/sourceMapGenerator.js:30:18)
    at workLoop (/Users/kylebaker/repos/beam-brush-unified/node_modules/metro/src/DeltaBundler/Serializers/sourceMapGenerator.js:40:22)
    at getSourceMapInfosImpl (/Users/kylebaker/repos/beam-brush-unified/node_modules/metro/src/DeltaBundler/Serializers/sourceMapGenerator.js:60:3)
    at sourceMapGenerator (/Users/kylebaker/repos/beam-brush-unified/node_modules/metro/src/DeltaBundler/Serializers/sourceMapGenerator.js:64:3)
    at sourceMapString (/Users/kylebaker/repos/beam-brush-unified/node_modules/metro/src/DeltaBundler/Serializers/sourceMapString.js:16:10)
    at /Users/kylebaker/repos/beam-brush-unified/node_modules/@sentry/react-native/dist/js/tools/vendor/metro/utils.js:42:21
    at Object.<anonymous> (/Users/kylebaker/repos/beam-brush-unified/node_modules/@sentry/react-native/dist/js/tools/sentryMetroSerializer.js:36:38)
    at Generator.next (<anonymous>)
    at /Users/kylebaker/repos/beam-brush-unified/node_modules/tslib/tslib.js:169:75
```

Figured it was something in `sentry` trying to call something incorrectly in metro, so it seems like it ended up being a tangential issue. Threw us for a loop for a minute because we didn't change anything with our sentry/metro versions but noticed renovate updated our lockfile which grabbed the newer versions of our `@types` packages, which included the `react-native: '*'` dependency which chained to grab all new versions of various `metro` packages. So setting `react-native` to our pinned version with a `resolutions` ended up being the fix for us. 

`node_modules` ends up having the correct `react-native` version but a lot of the dependencies from the incorrect version are what end up getting installed

**Comment by robhogan** [comment]
CC @krystofwoldrich for the above - this looks to be down to the way the Sentry integration tries to find the project's Metro without having an explicit dependency on it, and the hoisted one isn't guaranteed to be correct.

https://github.com/getsentry/sentry-react-native/blob/bfb067670e46d34f6d802392c219a22e6f03ed47/src/js/tools/vendor/metro/utils.ts#L28

**Comment by krystofwoldrich** [comment]
@kylebake Did I understand it correctly the issue was resolved by removing the multiple `react-native` versions? After that, was the `react-native: '*'` working?

**Comment by kylebake** [comment]
> @kylebake Did I understand it correctly the issue was resolved by removing the multiple `react-native` versions? After that, was the `react-native: '*'` working?

Yes, once we pinned `react-native` to our expected `0.71.7` version with a [`resolutions`](https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions/) in our package.json, everything was fine again. 

I'm not 100% clear on how everything was getting resolved but it seemed like when we had `react-native 0.73.1` in our lockfile, a lot of the dependencies (such as the metro packages) defined by that version were what ended up being installed in our node modules.

**Pull request merged**
